Add support for Spark SQL

wjoel opened the pull request:
Only tested with the Spark Thrift Server and Drill #2157, but might work with other Hive compatible SQL engines (Hive itself)



###### TODO 
-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).
-  [x] Add tests? If so, I might need some help. (edit: 2208 tests pass)

wjoel committed:
Add support for Spark SQL (Hive)

Only tested with the Spark Thrift Server, but might work with other
Hive compatible SQL engines (Hive itself and Drill #2157, etc.)
since it's using the Hive JDBC driver.

wjoel committed:
Improve formatting, remove some log messages used for debugging

wjoel committed:
Use NOW() for "now"

wjoel committed:
Reference JIRA issue for setReadOnly fix in FixedHiveConnection

salsakran commented:
You will *definitely* need to add tests =)

wjoel committed:
Don't use dashes in namespaces/driver name

Seems they stop the driver from being loaded.

wjoel committed:
Add support for Drill, improve SparkSQL/Hive support

wjoel committed:
Use our implementation of prepare-value

wjoel committed:
Fix DRILL-5383

wjoel committed:
Don't include schema name in field escapes

wjoel committed:
Copy and change unprepare, :timestamp "doesn't work for every DB"

Let Drill use :to_timestamp instead

wjoel committed:
Use hive-jdbc from Spark project

wjoel committed:
Replace dashes with underscores in table names

Dashes are not valid in table names in some SQL dialects, including
Spark SQL.

wjoel committed:
Specify features in Spark SQL and Hive drivers

wjoel committed:
Fix native queries (should not end with semicolons)

wjoel committed:
Rename db-with-some-cruft to db_with_some_cruft

wjoel committed:
Set timezone to UTC

wjoel committed:
Improve test coverage for Spark SQL, Hive

wjoel committed:
Many fixes to Drill and Spark, Drill now passes all tests

wjoel committed:
Make all tests pass for both Drill and Spark SQL

wjoel committed:
Remove Hive driver until we can get it to co-exist with Spark SQL.

All tests pass in this commit.

wjoel committed:
Enable foreign-keys for Drill and Spark SQL

The tests don't work since we can't create foreign keys with those
SQL engines, but they can perform joins if the foreign key has been
specified manually in the data model interface.

We should not include the schema name in the "ON" clause (it breaks
Drill). Note that the schema is already correctly specified in the
"FROM <schema>.table1" and "LEFT JOIN <schema>.table2 alias1" clauses,
so there should be no ambiguity in in "ON table1.field1 = alias.field2"

wjoel commented:
@salsakran, I can run `ENGINES=h2,sparksql,drill lein test` from [this commit](https://github.com/metabase/metabase/pull/4645/commits/408576444d85196cdcc2aad36ea0a8b1c4400c45) and get
```
Ran 2152 tests containing 2152 assertions in 142787 msecs
0 failures, 0 errors.
```

The "Enable foreign-keys for Drill and Spark SQL" does not pass the tests since those SQL engines do not support creating foreign keys, but can do joins just fine (in my limited testing) if a foreign key has been specified manually in the data model, so it would be a shame to disable that functionality entirely.

Please let me know how to proceed from here.

wjoel committed:
Undo damage caused by aggressive-indent, delete unused code

wjoel committed:
Add support for remaining tests in date_bucketing_test

wjoel committed:
Use functions from generic/ when possible, fix indentation

wjoel committed:
Support multiple databases in Spark SQL

wjoel committed:
Move back to using dashes in table, database definitions

All tests pass for Spark SQL and Drill with this commit.
Note that Drill must be configured with

"extractHeaders": true

in "formats" -> "csv" in the "dfs" storage plugin when
running the tests.

wjoel committed:
Enable foreign-keys for Spark SQL, Drill despite tests failing

... since they do work in actual usage, but creation of foreign
keys is not supported on the test (or database) side.

wjoel committed:
Use "Database name" rather than "Name" in Spark SQL queries

wjoel commented:
@salsakran, others: This is ready for a first review.

salsakran commented:
@wjoel can you sign our CLA? That's a blocking pre-requisite for us to take any PRs.

wjoel commented:
@salsakran, I've now signed your CLA.

camsaul commented on project.clj:
```diff
@@ -28,6 +28,18 @@
+                  :exclusions [org.mortbay.jetty/jetty
+                               org.eclipse.jetty.aggregate/jetty-all
+                               org.codehaus.jackson/jackson-xc]
+                  :classifier "standalone"]
```
can you keep the dependencies ordered alphabetically?

camsaul commented on project.clj:
```diff
@@ -40,7 +52,8 @@
                                net.sourceforge.nekohtml/nekohtml
-                               ring/ring-core]]
+                               ring/ring-core
+                               com.google.inject/guice]]
```
can you keep the exclusions ordered alphabetically as well?

camsaul commented on project.clj:
```diff
@@ -85,8 +99,11 @@
   :target-path "target/%s"
   :jvm-opts ["-server"                                                ; Run JVM in server mode as opposed to client -- see http://stackoverflow.com/questions/198577/real-differences-between-java-server-and-java-client for a good explanation of this
+             "-Xmx4g"
+             "-Duser.timezone=Etc/UTC"
```
what's the reasoning behind adding these settings? This would affect everyone who develops or runs a server via `lein`

Personally I prefer letting the JVM use as much memory as it wants on my dev machine. For example with 64 GB of RAM it often ends up using much more than 4GB. Most of the MB devs as well prefer sticking to our default system timezones of `US/Pacific` rather than UTC.

camsaul commented on src-java/FixedHiveConnection.java:
```diff
@@ -0,0 +1,52 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Metabase is distributed under the AGPL so we can't go around adding source files with huge comments explaining they're distributed under Apache

(right, @salsakran?)

camsaul commented on src-java/FixedHiveDriver.java:
```diff
@@ -0,0 +1,47 @@
+    public Connection connect(String url, Properties info) throws SQLException {
+        return acceptsURL(url) ? new FixedHiveConnection(url, info) : null;
+    }
+}
```
couldn't these classes just be written in Clojure? You can still extend classes and things like that. Take a look at https://clojure.org/reference/java_interop#_implementing_interfaces_and_extending_classes

camsaul commented on src-java/FixedHiveDriver.java:
```diff
@@ -0,0 +1,47 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Same issue. Mismatched licenses. @salsakran

camsaul commented on src/metabase/db/migrations.clj:
```diff
@@ -370,4 +370,4 @@
 (defmigration ^{:author "camsaul", :added "0.23.0"} drop-old-query-execution-table
   ;; DROP TABLE IF EXISTS should work on Postgres, MySQL, and H2
-  (jdbc/execute! (db/connection) [(format "DROP TABLE IF EXISTS %s;" ((db/quote-fn) "query_queryexecution"))]))
+  (jdbc/execute! (db/connection) [(format "DROP TABLE IF EXISTS %s" ((db/quote-fn) "query_queryexecution"))]))
```
why?

camsaul replied:
these statements won't even execute against Drill/etc., they only run against the application DB which is Postgres, MySQL, or H2

camsaul commented on src/metabase/driver/drill.clj:
```diff
@@ -0,0 +1,245 @@
+
+(defn- describe-database [driver database]
+  {:tables (with-open [conn (jdbc/get-connection (sql/db->jdbc-connection-spec database))]
+             (set (for [result (jdbc/query {:connection conn} ["select table_schema, table_name from INFORMATION_SCHEMA.`VIEWS` union select table_schema, table_name from INFORMATION_SCHEMA.`TABLES` where table_type='TABLE'"])]
```
You can't use the normal JDBC metadata functions to get this information?

camsaul commented on src/metabase/driver/drill.clj:
```diff
@@ -0,0 +1,245 @@
+                                                   (hx/literal (u/date->iso-8601 this))
+                                                   (hx/literal "YYYY-MM-dd''T''HH:mm:ss.SSSZ"))))))
+
+(defn- unprepare
```
Is this needed for Spark SQL/Drill? It's there for BigQuery because BigQuery isn't JDBC-based and doesn't have prepared statements. But if SparkSQL/Drill has a JDBC driver shouldn't prepared statements work normally?

camsaul commented on src/metabase/driver/generic_sql.clj:
```diff
@@ -220,7 +220,7 @@
 (defn- qualify+escape ^clojure.lang.Keyword
   ([table]       (hx/qualify-and-escape-dots (:schema table) (:name table)))
-  ([table field] (hx/qualify-and-escape-dots (:schema table) (:name table) (:name field))))
+  ([table field] (hx/qualify-and-escape-dots (:name table) (:name field))))
```
this is going to break things for people who use databases with multiple schemas.

camsaul commented on src/metabase/driver/generic_sql/query_processor.clj:
```diff
@@ -81,7 +81,7 @@
   Field
   (formatted [{:keys [schema-name table-name special-type field-name]}]
-    (let [field (keyword (hx/qualify-and-escape-dots schema-name table-name field-name))]
+    (let [field (keyword (hx/qualify-and-escape-dots table-name field-name))]
```
same issue. There's a reason for qualifying things with schema names. `schema_1.some_table` is different from `schema_2.some_table` but now we'll just have `some_table` and `some_table`.

If SparkSQL/Drill don't support schemas (`schema-name` is `nil`) the function takes care of that and leaves the schema out.

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,160 @@
+    (if-not (seq args)
+      sql
+      (recur (s/replace-first sql #"(?<!\?)\?(?!\?)" (unprepare-arg arg))
+             more-args))))
```
same thing as above. Is this needed for a JDBC-based driver?

wjoel replied:
It's slightly different, so it's not just copy-pasta. If there is a way to re-use the common parts of the protocol and modify only parts of it I'd be happy to do so, but I'd need your help to do so, because I don't see it. It has the same name, and the same purpose, but it's tweaked for this driver.

camsaul replied:
@wjoel there's actually a *much* easier way of doing this. You can define custom HoneySQL methods to handle conversion of a function to SQL. So what we can do here is just pass in a custom function to be our `iso-8601-fn` and add a custom handler for that:

```clojure
(require '[honeysql.format :as hformat])

;; custom method to convert a :hive-like-from-unixtime function call to SQL. 
;; hive-like-from-unixtime isn't an actual function, just a name we use so we can pass it to
;; `unprepare` as the `iso-8601-fn` 
(defmethod hformat/fn-handler "hive-like-from-unixtime" [_ datetime-literal]
  (hformat/to-sql
   (hsql/call :from_unixtime
     (hsql/call :unix_timestamp
       datetime-literal
       (hx/literal "yyyy-MM-dd\\\\'T\\\\'HH:mm:ss.SSS\\\\'Z\\\\'")))))

;; (using it)
(unprepare sql-and-args :iso-8601-fn :hive-like-from-unixtime)
```

Let me know if that makes sense. In general any time we're copying entire namespaces to change a single line of code we should look at other ways of doing things to reduce code duplication. We're aiming to keep the driver code as tight and clean as possible because things could quickly spiral out of control in terms of maintenance costs given the number of databases we support

wjoel replied:
Very nice, that seems to work beautifully.

camsaul commented on test/metabase/api/card_test.clj:
```diff
@@ -364,7 +364,7 @@
                   Card      [card              {:dataset_query {:database database-id
                                                                 :type     :native
-                                                                :native   {:query "SELECT COUNT(*) FROM CATEGORIES;"}}}]]
+                                                                :native   {:query "SELECT COUNT(*) FROM CATEGORIES"}}}]]
```
if the driver doesn't support terminating semicolons?

wjoel replied:
Correct, Drill (I think... it's possible it's Spark SQL) doesn't like them and they are not necessary.

camsaul commented on test/metabase/query_processor_test/date_bucketing_test.clj:
```diff
@@ -60,7 +60,7 @@
      ["2015-06-02T01:20:00.000Z" 1]
      ["2015-06-02T04:11:00.000Z" 1]])
-  (sad-toucan-incidents-with-bucketing :default))
+ (sad-toucan-incidents-with-bucketing :default))
```
I think the indentation was correct beforehand, why did you change this?

camsaul commented on test/metabase/query_processor_test/filter_test.clj:
```diff
@@ -76,12 +76,12 @@
+        (data/run-query places
+          (ql/filter (ql/!= $liked true))
+          (ql/order-by (ql/asc $id))))
+      rows (format-rows-by [int str ->bool] :format-nil-values)))
```
same thing, why change the indentation?

camsaul commented on test/metabase/test/data/drill.clj:
```diff
@@ -0,0 +1,187 @@
+      (catch Throwable e
+        (printf "Caught Exception: %s %s\n%s\n" (class e) (.getMessage e)
+                (with-out-str (.printStackTrace e)))
+        (throw e)))))
```
Why is this copied over from the generic-sql namespace? Can't you just use that implementation?

wjoel replied:
No, this is different, since it does not use prepared statements (Drill does not support prepared statements).

camsaul replied:
Wouldn't it be possible to write an implementation something like this?

```clojure
(defn sequentially-execute-sql! [driver context dbdef sql]
  (generic-sql/sequentially-execute-sql! driver context dbdef (unprepare sql))
```

(Just call the normal function with an unprepared version of the SQL)

I'm not comfortable with copying entire sections of code to change one little thing. It makes this PR way bigger than it needs to be and if we're not careful about keeping as much code shared as possible the driver code with spiral out of control in terms of complexity

wjoel replied:
Kind of. I renamed it to `sequentially-execute-sql-without-prepared-statements!` and managed to reuse some code from generic-sql, but there's still a little bit of repetition in `execute-without-prepared-statement!` that I wasn't able to get rid of. I hope it's good enough.

camsaul reviewed (changes requested):
Fix issues noted in comments

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,160 @@
+    (if-not (seq args)
+      sql
+      (recur (s/replace-first sql #"(?<!\?)\?(?!\?)" (unprepare-arg arg))
+             more-args))))
```
Also BTW there's a `metabase.driver.generic-sql.util.unprepare` namespace that has this protocol and functions in it already. So there should be no need to copy-pasta the entire namespace when you can just used the shared code

wjoel committed:
Fix describe-table, describe-database (we should use :db here)

wjoel committed:
Order dependencies, exclusions alphabetically

wjoel committed:
Add back semicolon to SQL statement that Drill doesn't care about

wjoel committed:
Use generic humanize-connection-error-message

wjoel committed:
Add schemas back in to qualify+escape

We'll reduce support for schemas/workspaces in Drill instead.

wjoel committed:
Add workaround for Drill's lack of schema-qualified field names

wjoel committed:
Undo accidental indentation changes

wjoel committed:
Explain why we don't use Drill's date_trunc_quarter

wjoel commented:
Hi @camsaul

Thank you for your review.

I've been able to fix most of your comments (I'll commit them in a day or two), but `qualify+escape` and `formatted` will be difficult for Drill. Drill does support schemas (of sorts), but it does not allow specification of the schema when referencing fields.
This is not a real limitation, because queries across schemas where tables have the same names and fields can use aliases, and I think Metabase should do so as well.

In other words, instead of doing this:
```SQL
select `dfs.tmp`.`tupac_sightings_cities`.`id` + `hive`.`tupac_sightings_cities`.`id`
from `dfs.tmp`.`tupac_sightings_cities`
left join `hive`.`tupac_sightings_cities`
on (`dfs.tmp`.`tupac_sightings_cities`.`id`=`hive`.`tupac_sightings_cities`.`id`)
where `dfs.tmp`.`tupac_sightings_cities`.`id` = 1;
```
which gives the error message `Error: VALIDATION ERROR: From line 3, column 11 to line 3, column 16: Duplicate relation name 'tupac_sightings_cities' in FROM clause`, we could/should be doing something like this:
```SQL
select a.`id` + b.`id`
from `dfs.tmp`.`tupac_sightings_cities` a
left join `hive`.`tupac_sightings_cities` b
on (a.`id`=b.`id`)
where a.`id` = 1;
```

I think most query generators use table aliases. Not only does it make Drill happy, it also shortens the query, which means fewer bytes sent across the wire and fewer bytes to parse for the database. For a toy example like this it's not much, but complicated queries with groups, joins and cast grow quickly with long table and schema names.
If this is not the right approach I'll have to find a way to ignore all schemas except the current one for Drill until we find a way to resolve this. The JDBC metadata does work for Drill (but not Spark SQL, see SPARK-9686), but then the query processor generates queries like this one:
```SQL
SELECT `dfs.tmp`.`places_cam_likes_places`.`liked` AS `liked`
FROM `dfs.tmp`.`places_cam_likes_places`
GROUP BY `dfs.tmp`.`places_cam_likes_places`.`liked`
ORDER BY `dfs.tmp`.`places_cam_likes_places`.`liked`
ASC LIMIT 301;
```
which fails with the error message `Error: VALIDATION ERROR: From line 3, column 10 to line 3, column 18: Table 'dfs.tmp' not found`
It can be rewritten to use a table alias, and then it works:
```SQL
SELECT a.`liked` AS `liked`
FROM `dfs.tmp`.`places_cam_likes_places` a
GROUP BY a.`liked`
ORDER BY a.`liked`
ASC LIMIT 301;
```

Leaving out the schema from the field references also works, but as you say, would cause issues with tables and fields with the same names but in different schemas:
```SQL
SELECT `places_cam_likes_places`.`liked` AS `liked`
FROM `dfs.tmp`.`places_cam_likes_places`
GROUP BY `places_cam_likes_places`.`liked`
ORDER BY `places_cam_likes_places`.`liked`
ASC LIMIT 301;
```

Are you planning to use table aliases in this way, or do you have any other ideas to work around Drill's lack of schema names in field references? I see table aliases are already supported for joins, so perhaps it's not too much work to use them in all queries.

Best regards,
  Joel

wjoel committed:
Use default implementations of describe-table, database for Drill

wjoel committed:
Rename Drill's implementation of unprepare

It is different from the generic-sql implementation because Date
and String are handled differently.

While the unprepare protocol serves the same purpose, it's different.

wjoel committed:
Explain need for describe-database and describe-table in Spark

wjoel committed:
Use gen-class for FixedHiveDriver and Connection

wjoel committed:
Remove schemas from qualify+escale and formatted to make Drill happy

This commit passes all tests.

wjoel committed:
Re-enable :foreign-keys for Drill and Spark SQL, break tests

salsakran commented:
Hey @camsaul -

There are some outstanding questions here. Can you take a look?

camsaul commented:
@wjoel using table aliases for all tables and fields would indeed be nice but it would probably be a pretty significant change to our query processor code to get it all working properly. 

However if you override the `field->identifier` method of `ISQLDriver` you can return an identifier that doesn't include the schema.

The normal implementation is basically
```clojure
(defn field->identifier [driver field]
  (apply hsql/qualify (field/qualified-name-components field)))
```

and returns something like `:schema.table.field`. So if you override that to drop the `schema.` part from the results the SQL that gets generated should look like what you need it to look like.

pygeek commented:
Curious to know if this is still being worked on? Seems stale.

wjoel commented:
I'm looking into adding support for table aliases. If it doesn't work out I'll go with the simplistic solution suggested above (but that would of course rule out joins between schemas).

pygeek commented:
@wjoel we could always start with the simple solution and iterate on that. Really looking forward to seeing this merged in 😃

wjoel committed:
Merge branch 'spark-sql' into sparksql-support

wjoel committed:
Use table aliases

wjoel committed:
Disable foreign keys in SparkSQL driver. All tests pass.

wjoel commented:
I've added support for table aliases, finally found an elegant way to do it (only requires changing 30 lines). I've only tested it with SparkSQL and H2 so far, but it looks promising. Hoping to push the changes next week after verifying it with Drill and PostgreSQL.

wjoel committed:
Add Drill driver

> ENGINES=h2,postgres,sparksql,drill lein test

`Ran 2846 tests containing 2846 assertions in 262912 msecs
0 failures, 0 errors.`

wjoel committed:
Merge branch 'master' of https://github.com/metabase/metabase into spark-sql

wjoel commented:
@salsakran @camsaul I'd like a new review on this, please. I believe all major concerns have been taken care of, and table aliasing is working fine from what I can tell (though it is a big change, so I hope you can help me test other databases). I'd love to get this merged - it's been a long time, and a lot of work.

```shell
ENGINES=h2,postgres,sparksql,drill lein test
...
Ran 2846 tests containing 2846 assertions in 262912 msecs
0 failures, 0 errors.
```

In order to support Hive in addition to Spark SQL it would be necessary to do some JAR shading. The Spark SQL dialect should be identical to Hive, but the JDBC driver is slightly different, and I don't have access to Hive, so someone else would have to fix that.

Remember to set `"extractHeader": true` in the `formats.csv` section of the `dfs` configuration for Drill before running the tests.

bholemt commented:
Is it possible to support options for HTTP transport mode also? The JDBC driver supports it.

wjoel committed:
Merge branch 'master' of https://github.com/metabase/metabase into spark-sql

wjoel committed:
Use schemas to represent different "databases" in SparkSQL

"databases" in SparkSQL are really closer to schemas in other databases.

wjoel commented:
@camsaul Could you please take another look at this? It's working well for me and the testers in #2157.

camsaul commented:
Hi @wjoel, we can take a look at this, and hopefully if everything looks good, ship it as part of our next release. I'm pretty busy today but I will take a look early next week

bholemt commented:
@wjoel I merged your branch with a fork I took from the latest release of metabase. It works great except one small thing. In the latest version there is a feature that lets you select a question as a table. However, questions I created on spark sql db are not shown here in "Saved Questions" menu (on the new questions page). Is it possible to address this? A big thanks again for your contribution.

wjoel committed:
Merge branch 'master' of github.com:wjoel/metabase into spark-sql

camsaul commented on project.clj:
```diff
@@ -29,6 +29,17 @@
+                  :exclusions [org.codehaus.jackson/jackson-xc
+                               org.eclipse.jetty.aggregate/jetty-all
+                               org.mortbay.jetty/jetty]
+                  :classifier "standalone"]
```
can you keep the dependencies in alphabetical order? That makes it easier to see what we're currently using/find one you're looking for

camsaul replied:
also https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#dependencies-should-have-comment

wjoel replied:
Done (passing eastwood)

camsaul commented on project.clj:
```diff
@@ -39,9 +50,10 @@
                  [com.cemerick/friend "0.2.3"                         ; auth library
-                  :exclusions [commons-codec
-                               org.apache.httpcomponents/httpclient
+                  :exclusions [com.google.inject/guice
```
just curious, what's the reason behind this change?

wjoel replied:
I was able to remove it

camsaul commented on project.clj:
```diff
@@ -90,7 +103,8 @@
                  [ring/ring-core "1.6.0"]
-                 [ring/ring-jetty-adapter "1.6.0"]                    ; Ring adapter using Jetty webserver (used to run a Ring server for unit tests)
+                 [ring/ring-jetty-adapter "1.6.0"                     ; Ring adapter using Jetty webserver (used to run a Ring server for unit tests)
+                  :exclusions [commons-io]]
```
same. Just trying to understand the reasoning behind this change

wjoel replied:
I was able to remove it

camsaul commented on src/metabase/driver/FixedHiveConnection.clj:
```diff
@@ -0,0 +1,22 @@
+   :init init
+   :constructors {[String java.util.Properties] [String java.util.Properties]}))
+
+(set! *warn-on-reflection* true)
```
we don't want this in the codebase because it will lead to tons of warning messages for things in our dependencies making it needlessly verbose for people just trying to use Metabase without doing development on it

wjoel replied:
Done

camsaul commented on src/metabase/driver/FixedHiveDriver.clj:
```diff
@@ -0,0 +1,17 @@
+   :prefix "driver-"
+   :constructors {[] []}))
+
+(set! *warn-on-reflection* true)
```
same thing, we need to take this out

wjoel replied:
Done

camsaul commented on src/metabase/driver/bigquery.clj:
```diff
@@ -393,8 +393,8 @@
+  "An implementation of `apply-aggregation` that just hands off to the normal Generic SQL implementation, but calls `deduplicate-select-aliases` on the results."
   [driver honeysql-form query]
   (-> (sqlqp/apply-aggregation driver honeysql-form query)
       (update :select deduplicate-select-aliases)))
```
if we're going to change this so it's used by multiple drivers we need to move it out of the bigquery namespace. We can move it into `metabase.driver.generic-sql.query-processor`, and put it after the default implementation of `apply-aggregation` and give it a name so it's clear to everyone that it's an alternative implementation. `apply-aggregation-deduplicate-select-aliases` is a bit long but describes exactly what it does

wjoel replied:
Done

camsaul commented on src/metabase/driver/drill.clj:
```diff
@@ -0,0 +1,196 @@
+            [metabase.driver.bigquery :as bigquery]
+            [metabase.util :as u]
+            [metabase.util.honeysql-extensions :as hx]
+            [metabase.query-processor.util :as qputil])
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#keep-requires-ordered

wjoel replied:
Done

camsaul commented on src/metabase/driver/drill.clj:
```diff
@@ -0,0 +1,196 @@
+(defn- trunc-with-format [format-str expr]
+  (str-to-date format-str (date-format format-str expr)))
+
+(defn date [unit expr]
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-things-private

wjoel replied:
Done

camsaul commented on src/metabase/driver/drill.clj:
```diff
@@ -0,0 +1,196 @@
+                         :quote-style (constantly :mysql)
+                         :current-datetime-fn (u/drop-first-arg (constantly hive-like/now))
+                         :string-length-fn (u/drop-first-arg hive-like/string-length-fn)
+                         :unix-timestamp->timestamp (u/drop-first-arg hive-like/unix-timestamp->timestamp)}))
```
This indentation is a bit weird check out the way the same sorts of things are indented in other drivers. e.g. `driver/IDriver` and the stuff below it should be indented 2 spaces instead of being even with `DrillDriver` at the top

wjoel replied:
Done, I think - I'm finding the indentation a bit challenging. I'm using Emacs but the indentation settings that are applied from your project do not seem to match current practice.

camsaul replied:
We've added `:style/indent` metadata to almost all of the functions where we want things to be indented a certain way, so CIDER should take care of things for you when you reindent the form

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,173 @@
+(defn- trunc-with-format [format-str expr]
+  (str-to-date format-str (date-format format-str expr)))
+
+(defn date [unit expr]
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-things-private

wjoel replied:
Done

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,173 @@
+   [metabase.driver.generic-sql :as sql]
+   [metabase.util :as u]
+   [metabase.util.honeysql-extensions :as hx]
+   [metabase.query-processor.util :as qputil]
```
these are almost all in perfect alphabetical order (thanks!) except for this query-processor entry https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#keep-requires-ordered

wjoel replied:
Done

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,173 @@
+   :boolean :type/Boolean
+   :binary :type/*})
+
+(def ^:const now (hsql/raw "NOW()"))
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#public-vars-must-have-a-useful-docstring

wjoel replied:
Done

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,173 @@
+  (apply hsql/qualify (qualified-name-components field)))
+
+;; copied from the Presto driver, except using mysql quoting style
+(defn apply-page [honeysql-query {{:keys [items page]} :page}]
```
Same thing for this function and the other ones in this namespace, they either should be private because they're not being used elsewhere, or the need to have a docstring explaining what they're for and how/why someone else might want to use them

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,173 @@
+  (try (f)
+       (catch java.sql.SQLException e
+         (log/error (jdbc/print-sql-exception-chain e))
+         (throw (Exception. (exception->nice-error-message e))))))
```
This looks like an exact copy-paste of the code in `metabase.driver.generic-sql.query-processor`, any reason why you can't just use that implementation?

wjoel replied:
Done (I made it public)

camsaul commented on src/metabase/driver/sparksql.clj:
```diff
@@ -0,0 +1,146 @@
+            [metabase.driver.hive-like :as hive-like]
+            [metabase.util :as u]
+            [metabase.util.honeysql-extensions :as hx]
+            [metabase.query-processor.util :as qputil]))
```
sorting

wjoel replied:
Done

camsaul commented on src/metabase/driver/sparksql.clj:
```diff
@@ -0,0 +1,146 @@
+  (loop []
+    (when-let [driver (try
+                        (java.sql.DriverManager/getDriver "jdbc:hive2://localhost:10000")
+                        (catch java.sql.SQLException e
```
https://github.com/bbatsov/clojure-style-guide#underscore-for-unused-bindings

wjoel replied:
Done

camsaul commented on src/metabase/driver/sparksql.clj:
```diff
@@ -0,0 +1,146 @@
+  ;; manually register our FixedHiveDriver with java.sql.DriverManager
+  ;; here and make sure it's the only driver returned for jdbc:hive2
+  ;; Class/forName and invokeConstructor is required to make this compile,
+  ;; but it may be possible to solve this with the right project.clj magic
```
Since you added these namespaces to the AOT list in `project.clj` do you still need to do all of this?

wjoel replied:
Yes, it's still necessary. The parent class for the `FixedHiveDriver` registers itself with JDBC for the `jdbc:hive2` protocol, and we must remove it in order for `FixedHiveDriver` (which is compatible with Spark SQL and works around some critical bugs, as documented).

camsaul commented on src/metabase/driver/sparksql.clj:
```diff
@@ -0,0 +1,146 @@
+    (s/replace s #"-" "_")))
+
+;; workaround for SPARK-9686 Spark Thrift server doesn't return correct JDBC metadata
+(defn describe-database [driver {:keys [details] :as database}]
```
Same thing as mentioned with other functions, we should either make these `^:private` if they're not used anywhere else or add docstrings explaining what the difference is between them and the normal implementation of `describe-database`

camsaul replied:
(Also applies to other fns in this namespace)

wjoel replied:
Done

camsaul commented on src/metabase/driver/sparksql.clj:
```diff
@@ -0,0 +1,146 @@
+                         :quote-style (constantly :mysql)
+                         :current-datetime-fn (u/drop-first-arg (constantly hive-like/now))
+                         :string-length-fn (u/drop-first-arg hive-like/string-length-fn)
+                         :unix-timestamp->timestamp (u/drop-first-arg hive-like/unix-timestamp->timestamp)}))
```
indentation

wjoel replied:
Done

camsaul commented on test/metabase/driver/generic_sql_test.clj:
```diff
@@ -18,7 +18,7 @@
   (delay (set (for [engine datasets/all-valid-engines
                     :let   [driver (driver/engine->driver engine)]
-                    :when  (not (contains? #{:bigquery :presto} engine))                 ; bigquery and presto don't use the generic sql implementations of things like `field-avg-length`
+                    :when  (not (contains? #{:bigquery :presto :sparksql} engine))       ; bigquery and presto don't use the generic sql implementations of things like `field-avg-length`
```
if you're going to change the code here you should update the comment as well. Something like `bigquery, presto, etc. don't use...` (changing just the beginning) would suffice

https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-sure-to-update-comments

wjoel replied:
Done

camsaul commented on test/metabase/query_processor_test/nested_queries_test.clj:
```diff
@@ -235,10 +235,10 @@
+                   "SELECT STDDEV(\"t1\".\"ID\") AS \"stddev\", \"t1\".\"PRICE\" AS \"PRICE\" "
+                   "FROM \"PUBLIC\".\"VENUES\" \"t1\" "
+                   "GROUP BY \"t1\".\"PRICE\" "
+                   "ORDER BY \"stddev\" DESC, \"t1\".\"PRICE\" ASC"
```
I'm not sure how I feel about changing a test shared across all of our DBs just to satisfy one new driver. I'd be surprised if this didn't end up breaking tests for other drivers

wjoel replied:
It's not just one driver though, this change affects all drivers. We should make sure it works with all of them, so I think changing the test makes sense. Please do help with testing other drivers (I've tested H2, Spark SQL, Drill, Postgres).

camsaul replied:
Ended up breaking BigQuery 😭

camsaul replied:
fixed in #7323

camsaul commented on test/metabase/test/data/drill.clj:
```diff
@@ -0,0 +1,177 @@
+                                [interface :as i])
+            [metabase.util :as u]
+            [metabase.util.honeysql-extensions :as hx]
+            [clojure.java.io :as io])
```
require sorting

wjoel replied:
Done

camsaul replied:
it's still out of order

camsaul replied:
If you're using emacs check out the package `clj-refactor`. If you `(setq cljr-favor-prefix-notation t)` and run the command `clr-clean-ns` it will automatically fix namespace require sorting for you. That's what myself and the other core metabase devs use

camsaul commented on src/metabase/driver/generic_sql/query_processor.clj:
```diff
@@ -97,8 +114,9 @@
   Field
-  (formatted [{:keys [schema-name table-name special-type field-name]}]
-    (let [field (keyword (hx/qualify-and-escape-dots schema-name table-name field-name))]
+  (formatted [unresolved-field]
```
What do you mean be calling this a "unresolved" field?

wjoel replied:
It needs to be resolved to an alias. Makes sense to me, but I can change it if you like.

camsaul replied:
It might make sense to think of a slightly different name. We use "unresolved Field" in an earlier stage in the query processor to refer to Fields that are referenced in the query that we haven't fetched from the DB yet. So this could end up getting a little confusing.

camsaul commented on src/metabase/driver/generic_sql/query_processor.clj:
```diff
@@ -32,6 +32,8 @@
   Each nested query increments this counter by 1."
   0)
 
+(def source-table-alias "t1")
```
If I understand this correctly, you're changing the code to give every table in a source query an alias of `t1`. What happens if something with a `JOIN` is used as a source query? Wouldn't this change break those queries?

wjoel replied:
Yes, this is a big change, and it needs to be tested with other drivers. However, my testing indicates that joins are fine, and even enabling `:nested-queries` for Drill and Spark SQL still passes all tests.
Even if we `JOIN` we should get (at worst) something like `select a from sometable t1 join (select b, c from othertable t1) _metabase_join_alias where (t1.a == _metabase_join_alias.b)` which is still ok.

At least, that's what I think. Again, this needs to be tested with all drivers (most of which I do not have access to).

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,173 @@
+            (h/where [:> :__rownum__ offset])
+            (h/limit items))))))
+
+;; lots of copy-paste here. consider making some of those non-private instead.
```
Yeah, please just make things public if you need to use them in some other namespace instead of copying them verbatim. We should try to avoid duplicating code here. A bug in one function becomes a bug every where that function gets copy-pastaed to.

wjoel replied:
Done

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,173 @@
+  (apply hsql/qualify (qualified-name-components field)))
+
+;; copied from the Presto driver, except using mysql quoting style
+(defn apply-page [honeysql-query {{:keys [items page]} :page}]
```
Same thing, copy-pasta could be avoided here. We can just move this into `metabase.driver.generic-sql.query-processor` and make it an official alternative implementation of `apply-page`. `quoting-style` can be an optional param. Then both `presto` and `hive-like` can use this implementation.

We've been following this sort of pattern where possible in the driver code. For example in `metabase.driver.generic-sql` we offer two different implementations of `active-tables`, so a driver can pick the one most appropriate for them (the differences and use cases are explained in their docstrings).

wjoel replied:
Done (I named it `apply-page-using-row-number-for-offset`)

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,173 @@
+(defn date-interval [unit amount]
+  (hsql/raw (format "(NOW() + INTERVAL '%d' %s)" (int amount) (name unit))))
+
+(defn string-length-fn [field-key]
```
As with other namespaces, functions in this one should be private if possible otherwise they need to be documented

camsaul replied:
I don't know if you saw it [in the Developers' Guide](https://github.com/metabase/metabase/blob/master/docs/developers-guide.md#run-the-linters) but you can and should run the liners locally:

```bash
lein eastwood && lein bikeshed && lein docstring-checker && ./bin/reflection-linter
```

That will catch these sorts of things and save us both some back-and-forth on the PR

wjoel replied:
Sorry, I may have missed that one, or it was added after the last time I looked at the Developer's Guide. My latest patch set passes the linters.

camsaul commented on test/metabase/test/data/drill.clj:
```diff
@@ -0,0 +1,177 @@
+   that don't support executing multiple statements at once.
+
+   Since there are some cases were you might want to execute compound statements without splitting, an upside-down ampersand (`⅋`) is understood as an
+   \"escaped\" semicolon in the resulting SQL statement."
```
Also at any rate worth repeating but if you copy-paste a function and change what it does you need to change the docstring too. Docstrings that are out-of-date or wrong are worse than no docstrings at all. Since other people might be looking at this driver as an example of how they write their own we want to make their lives a bit easier and make sure the code is clean and readable.

https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-sure-to-update-comments

wjoel replied:
Done

camsaul commented on test/metabase/test/data/sparksql.clj:
```diff
@@ -0,0 +1,120 @@
+            (metabase.test.data [generic-sql :as generic]
+                                [interface :as i])
+            [metabase.util :as u]
+            (honeysql [core :as hsql]
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#keep-requires-ordered

wjoel replied:
Done

camsaul reviewed (changes requested):
Thanks for all your hard work so far @wjoel. There's a few more things we need to do before we merge this:

*  There are a bunch of places with copied code that I think we can rework to avoid duplication. Since other people will be looking at this driver as an example of how to write their own, it's important we try to keep the code down to just what's necessary so people don't get confused or overwhelmed. Plus, with the number of databases we support, we need to try to minimize duplication so a bug in one place doesn't become a bug in 12. (In other words, duplicated code is harder for us to maintain.) 

    For a few of these I tried to post code or other instructions for how we can reuse existing functions, but let me know if you have any questions, I am happy to help get things cleaned up
*  There are a bunch of minor style things like namespace `:require` sorting still in this PR. I mentioned them last time I reviewed the code but it looks like there's still a few places to fix. I'd highly recommend taking a look at our [Style Guide](https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide) 
*  There's a bunch of things that wouldn't pass the linter. Make sure you run it locally with 
    ```
    lein eastwood && lein bikeshed && lein docstring-checker && ./bin/reflection-linter
    ```

We'd love to ship this as part of our next release, so let me know if you need help making any of these changes to get things wrapped up.

wjoel committed:
Address review comments

wjoel commented on src/metabase/driver/generic_sql/query_processor.clj:
```diff
@@ -79,6 +81,21 @@
   (formatted [this]
     "Return an appropriate HoneySQL form for an object."))
 
+(defn resolve-table-alias [{:keys [schema-name table-name special-type field-name] :as field}]
```
Done

wjoel reviewed (commented):
Thanks for the review, please take a look at the new patch set

camsaul commented on project.clj:
```diff
@@ -39,9 +39,7 @@
                  [com.amazon.redshift/redshift-jdbc41 "1.2.8.1005"]   ; Redshift JDBC driver
                  [com.cemerick/friend "0.2.3"                         ; auth library
-                  :exclusions [commons-codec
-                               org.apache.httpcomponents/httpclient
```
Just so I understand what's going on here. Do the drivers depend on having `org.apache.httpcomponents/httpclient` around?

camsaul commented on test/metabase/test/data/sparksql.clj:
```diff
@@ -0,0 +1,120 @@
+                 (merge generic/IDriverTestExtensionsMixin
+                        {:database->connection-details (u/drop-first-arg database->connection-details)
+                         :default-schema               (constantly "test_data")
+                         :engine                       (constantly :sparksql)}))
```
indentation is still weird here

camsaul commented on test/metabase/test/data/sparksql.clj:
```diff
@@ -0,0 +1,120 @@
+      (try
+        (do
+          (.setAutoCommit conn false)
+          (jdbc/execute! {:connection conn} sql+args {:transaction? false}))
```
You don't need to wrap multiple statements in a `try` form inside a `do` form.

You can just do

```clojure
(try
  (.setAutoCommit conn false)
  (jdbc/execute! {:connection conn} sql+args {:transaction? false}))
  (catch ...))
```

camsaul commented:
@wjoel could you help me get the drivers set up to run in CI? Specially a little bit of scripting to install and start the servers.

Our CI script lives here: https://github.com/metabase/metabase/blob/master/bin/ci

We need a couple of functions called something like `install-spark()` and `install-drill()` and then need to tell one of our nodes to run those and add `spark,drill` to the list. If you could help me write those functions to start the DB processes I can do the rest if that works.

Thanks!
Cam

camsaul commented on src/metabase/driver/drill.clj:
```diff
@@ -0,0 +1,184 @@
+                                        (hx/literal (u/date->iso-8601 this))
+                                        (hx/literal "YYYY-MM-dd''T''HH:mm:ss.SSSZ")))
+  Number (prepare-value [this] this)
+  Object (prepare-value [this] (throw (Exception. (format "Don't know how to prepare value %s %s" (class this) this)))))
```
Copying this entire protocol and set of implementations over from the BigQuery to only change one line sounds like a like of code duplication. I'm happy to fix this and rework this in a way that avoids copy pasta but it's something I want to address before merging

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,149 @@
+               (qualified-name-components parent)
+               (let [{table-name :name, schema :schema} (db/select-one ['Table :name :schema], :id table-id)]
+                 [table-name])))
+        field-name))
```
It looks like this implementation is the same as the one in `metabase.models.field`, but you took out the part that optionally prepends `schema` (it looks like you're still fetching it from the DB, however). Is this because Hive DBs don't have schemas? If that's the case, just make sure the Tables are getting stored in the DB with a `nil` `schema`. If you take a look at the original function it only includes schema as a "name component" if it's truthy. So you should still be able to use that implementation.

wjoel replied:
I think it would be ok to remove this entirely, but doing so breaks three tests in `metabase.query-processor.middleware.parameters.sql-test` for Spark SQL and Drill (everything else passes just fine). The tests are on lines 447, 468, 479. I suspect that those three tests should be updated to use aliasing as well, or something like it, but I'm not sure what they're testing exactly.

camsaul commented on src/metabase/driver/hive_like.clj:
```diff
@@ -0,0 +1,149 @@
+(defn run-query-without-timezone
+  "Runs the given query without trying to set a timezone"
+  [driver settings connection query]
+  (run-query query connection))
```
It looks like these are copied over from the implementations in `generic-sql.query-processor` but you just took the timezone stuff out. Those functions are supposed to work for databases that don't support setting a timezone (defined by whether they declare the `:set-timezone` feature). What was the issue you were having with them? If you couldn't get them to work it's probably better we tweak them so we can reuse them instead of copying them to change one tiny thing

wjoel replied:
The big issue is the use of transactions, not the timezone thing. I tried modifying the code to take `transaction?` as an argument, but I didn't like the result. I think it should probably be in the driver features.

camsaul commented on test/metabase/test/data/drill.clj:
```diff
@@ -0,0 +1,188 @@
+      (let [sql (s/replace sql #";+" ";")]
+        (try
+          (with-open [sql-statement (.createStatement conn)]
+            (.execute sql-statement sql))
```
It looks like this part is the only part that's different from the `default-execute-sql!` implementation. It would be a lot nicer just to tweak the original function for flexibility rather than copying the whole thing over to change one thing.

One thing we could do is let it take an optional `execute` function the way `sequentially-execute-sql!` does:

```clojure
(defn- jdbc-execute! [db-spec sql]
  (jdbc/execute! db-spec [sql] {:transaction? false, :multi? true}))

(defn default-execute-sql! [driver context dbdef sql & {:keys [execute]
                                                        :or {execute jdbc-execute!}}]
  (let [sql (some-> sql s/trim)]
    (when (and (seq sql)
               ;; make sure SQL isn't just semicolons
               (not (s/blank? (s/replace sql #";" ""))))
      ;; Remove excess semicolons, otherwise snippy DBs like Oracle will barf
      (let [sql (s/replace sql #";+" ";")]
        (try
          (execute db-spec sql)
          (catch SQLException e
            (println "Error executing SQL:" sql)
            (printf "Caught SQLException:\n%s\n"
                    (with-out-str (jdbc/print-sql-exception-chain e)))
            (throw e))
          (catch Throwable e
            (println "Error executing SQL:" sql)
            (printf "Caught Exception: %s %s\n%s\n" (class e) (.getMessage e)
                    (with-out-str (.printStackTrace e)))
            (throw e)))))))
```

The this whole implementation here could look like

```clojure
(def ^:dynamic ^:private *connection* nil)

;; workaround for DRILL-5136 (can't use prepared statements for DDL)
;; instead of prepared statements, we use plain statements.
;; since execute in the jdbc driver uses prepared statements by default,
;; we use .createStatement and .execute explicitly.
(def ^:private execute-without-prepared-statement! [driver context dbdef sql]
  (generic/default-execute-sql! driver context dbdef sql
                                :execute (fn [_ sql]
                                           (with-open [sql-statement (.createStatement *connection*)]
                                             (.execute sql-statement sql)))))

(defn sequentially-execute-sql-without-prepared-statements! [driver context dbdef sql]
  (with-open [conn (jdbc/get-connection (generic/database->spec driver context dbdef))]
    (binding [*connection* conn]
      (generic/sequentially-execute-sql! driver context dbdef sql
                                         :execute (partial execute-without-prepared-statement! conn)))))
```

That would save us 30 lines of copy-pasta or so

camsaul replied:
Alternatively even just abstracting out the error handling and logging so it can be reused would be a huge win 

Keeping code duplication to a minimum here is a good thing because if we can make the normal driver implementations more flexible/easier to assemble together it will make maintainence and writing future drivers easier

camsaul commented:
@wjoel (or anyone else) are you able to use these drivers with Java 9? I really would like to see this merged so I've been trying to work on some things on my end but I can't seem to run the tests:

```
    threw: class java.lang.NoClassDefFoundError - sun/misc/VM
           oadd.org.apache.drill.common.config.DrillConfig$<clinit> (DrillConfig.java:51)
           org.apache.drill.jdbc.impl.DrillConnectionImpl$<init> (DrillConnectionImpl.java:142)
           org.apache.drill.jdbc.impl.DrillJdbc41Factory$newDrillConnection (DrillJdbc41Factory.java:72)
           org.apache.drill.jdbc.impl.DrillFactory$newConnection (DrillFactory.java:69)
           oadd.org.apache.calcite.avatica.UnregisteredDriver$connect (UnregisteredDriver.java:143)
           org.apache.drill.jdbc.Driver$connect (Driver.java:72)
           java.sql.DriverManager$getConnection (DriverManager.java:678)
           java.sql.DriverManager$getConnection (DriverManager.java:190)
           clojure.java.jdbc$get_connection$invokeStatic (jdbc.clj:358)
           on (jdbc.clj:225)
           clojure.java.jdbc$get_connection$invokeStatic (jdbc.clj:277)
           metabase.test.data.drill$sequentially_execute_sql_without_prepared_statements_BANG_$invokeStatic (drill.clj:102)
```

It sounds like one of the dependencies is using `sun.misc.VM` which is an [undocumented/unsupported internal class that was moved or removed for Java 9](https://stackoverflow.com/questions/46640462/noclassdeffounderror-sun-misc-vm-from-ehcache-on-migrating-to-java-9-during-ru).

wjoel committed:
Address review comments, round 2

camsaul commented:
@wjoel let me know if you can help me figure out how to get these DBs running locally, that's one of the last big parts left needed for getting this merged. If you can help me with that I can probably manage the rest of the tweaks needed to get this merged

wjoel commented:
I haven't tried Java 9, no. Some quick googling doesn't show anything promising for Drill and Java 9.

To run locally:
Download Drill, unpack it and start it in embedded mode: `cd apache-drill-1.10.0/bin && ./drill-embedded`, then go to `http://localhost:8047/storage` and click "Update" on `dfs`, and add `"extractHeader": true` in `formats.csv`. It should be possible to script this with curl, but I haven't looked into it. For that matter, it might be possible to supply this as a configuration file, but I haven't tried.

I've only tried Drill 1.10. Now Drill 1.11 is out, upgrading is probably not an issue.

For Spark SQL:
Download Spark 2.1.2, unpack it and run `cd spark-2.1.2-bin-hadoop2.7 && java -Duser.timezone=Etc/UTC -Xmx1g -cp "$(pwd)/conf:$(pwd)/jars/*" org.apache.spark.deploy.SparkSubmit --master local[8] --conf spark.executor.extraJavaOptions=-Duser.timezone=Etc/UTC --conf spark.cores.max=16 --class org.apache.spark.sql.hive.thriftserver.HiveThriftServer2 --name "Thrift JDBC/ODBC Server" --executor-memory 10g spark-internal`
Same thing here, Spark 2.2.0 has since been released, but I don't expect any difficulties when upgrading. I just didn't want to add more work for this first commit.

Note that in order for the tests to pass, `:foreign-keys` must be disabled in the features for both Spark SQL and Drill, even if they both support foreign keys. This is because the tests for that feature also attempts to *create* foreign keys, which they do not support. However, if foreign keys are added manually in the Metabase database model, the joins (as far as I, and others, can tell from using it) on foreign keys works just fine.

You might want to change the tests to use a Metabase data model that has foreign keys pre-defined, leave this (very useful) feature disabled, or disable the tests for those drivers but leave the feature enabled.

wjoel commented:
@camsaul please let me know if that helped, and what the required steps would be to get this merged for 0.28. I would like to finish this project.

camsaul commented:
@wjoel yes, that helped a lot. I'm also aiming to get this merged into 0.28, but we can't merge it if it breaks Java 9 support. So I'm really hoping whatever upstream dependency is breaking that gets updated

wjoel commented:
Great! I'll try to give Java 9 a shot on Sunday, but next week looks more likely.

wjoel commented:
Spark SQL is ok with Java 9, but Drill doesn't work due its use of `sun.misc.VM` which has been moved to `jdk.internal.misc.VM`.

See https://github.com/apache/drill/blob/master/common/src/main/java/org/apache/drill/common/config/DrillConfig.java#L51

salsakran commented:
@wjoel If that's the case, would it make sense to split off drill support into a separate PR and just merge spark for 0.28?

wjoel commented:
@salsakran Sure, we can do that, but is feature parity between Java 8 and Java 9 a must have for Metabase? Could just disable the driver if running on Java 8, for example.

salsakran commented:
@wjoel yeah, our fundamental stance is that installing and administering Metabase should be as close to trivial as we can make it. We're already stretching that promise pretty thin with having people stuff things on classpaths for proprietary drivers (Vertica, Oracle), and dealing with beanstalk, etc, but I want to make things better and not worse. 

The entire promise of the jvm was write-once, run-anywhere, not "some shit works some shit doesn't, do a java -version and find fun surprises". My read is some of this java 8/9 incompatibility in the underlying libraries will resolve itself in time. Am I fundamentally wrong on that?

wjoel commented:
@salsakran Yeah, it's unfortunate that some performance focused libraries have been using internal JDK APIs for such a long time. Java 9 will kind of fix this by forcing libraries to change, but it does so by intentionally breaking them. I attempted to fix the Drill JDBC driver by using reflection to get the proper class for getting the value of `maxDirectMemory`, but that was only the first issue.

New versions of libraries that are maintained will eventually get support for Java 9+, yes. They kind of have to. However, this could take some time. As for Drill, I haven't been able to find any open issues or email discussions to fix this, so I don't know if their community is unaware, doesn't care, or is already working on fixing it but without bothering to announce it. I'm not actually using Drill myself, so I'm not going to get involved to try and find out, just seems like a bit of a shame to not be able to get Drill support merged since probably only a tiny fraction of Metabase users are using Java 9.

Still, if Drill users want Metabase to support Drill and you require that it supports Java 9, let them fix their JDBC library first. I'd rather finish this for now, even if it ends up being just Spark SQL (which I think is far more important, anyway). I'll try to get it done after the holidays.

salsakran commented:
Yeah, I'd say I'm a huge +1 on just getting spark in and merged. 

This is perhaps just a reflection of my own bubble, but I know tons of people using Spark and no-one using Drill.

wjoel committed:
Merge branch 'master' of github.com:wjoel/metabase into spark-sql

wjoel committed:
Remove Drill driver since it doesn't work with Java 9

wjoel committed:
Fix Java 9 compatibility for Spark SQL

wjoel committed:
Sync with master

wjoel committed:
Merge branch 'master' of https://github.com/metabase/metabase into spark-sql

wjoel commented:
Drill is gone, all 2812 tests pass on `java version "9.0.1"` and `openjdk version "1.8.0_151"` with `ENGINES=h2,postgres,sparksql lein test`.

wjoel committed:
Remove mentions of Drill from tests

cgivre commented:
Hi @wjoel , 
Was the issue with Drill that it doesn't support Java 9?

wjoel commented:
@cgivre the first issue was the use of `sun.misc.VM` in https://github.com/apache/drill/blob/master/common/src/main/java/org/apache/drill/common/config/DrillConfig.java#L51 and after I removed that there were some more issues with `drill/exec/memory/base/src/main/java/io/netty/buffer`, which is where I gave up.

petamex commented:
Is there something new on the topic? We really really would love to have Spark SQL support in Metabase.

wjoel commented:
Hi @petamex, good question. Nothing new from my side. It's been almost a year since I first submitted this PR, and I did a lot of work on this so of course I'd very much like to see it merged, but I am no longer working with Metabase or Clojure and I'm involved in other projects now.

I'm willing to do one last rebase if the Metabase maintainers commit to actually helping me get this into master, but then/otherwise I'm likely to just move on and leave this for someone else to maybe finish up one day.

salsakran commented:
@wjoel First off thank you so much for all the work that's gone into this PR. We really appreciate both your desire to help out, actually getting the first PR done and then taking all of our feedback and updating the PR. 

It's unfortunate we weren't able to merge this in November due to the Java 9 issue. We also apologize for how slow the code review process has been. 

Some RealTalk that has nothing to do with your own behaviour but hopefully will explain our own behaviour:
We've had a hard time with some community contributed database drivers over the years. When the first rush of adding a big new feature has faded, the core team is left with supporting the driver for ever. This is especially difficult when someone on the core team or an active contributor isn't running the specific database in production. Bugs pop up, database jdbc drivers get updated and break things, and in general time that we could spend working on our ever increasing backlog of features and bug fixes gets eaten up. As a result of a couple of these episodes, we've responded by making the barrier to a first merge higher. I really regret some of the delays between these stages, however, we've had to be a little OCD given the realities of how much work maintaining drivers has been. I don't say any of this as a reflection on anything you've done, but rather how we've had to adjust based on the behaviour of others. 

@camsaul @senior can you take a look at this PR? Is it gtg other than the rebase?

camsaul commented:
Yeah I think it's gtg we just have to get CI set up

lucasloami commented:
Are there any updates in this issue?

wjoel committed:
Merge branch 'master' of https://github.com/metabase/metabase into spark-sql

wjoel committed:
Adjust or fix new tests for Spark SQL

wjoel committed:
Enable more features and disable foreign-keys in testing like Presto

wjoel commented:
@salsakran @camsaul I've rebased and updated this, hope we can get it merged soon.

```
LEIN_JVM_OPTS="-Xmx2g" ENGINES=h2,postgres,sparksql lein test

Ran 3101 tests containing 3101 assertions in 136445 msecs
0 failures, 0 errors.
```

salsakran commented:
Thanks @wjoel !

the test suite passes w/ spark 2.3.0 when I just ran it. 
~3 minutes on a macbook.

salsakran commented:
@camsaul @senior @tlrobinson which ci slot should this go into?

camsaul commented:
@salsakran node 0 since that's the backend test one that currently finishes soonest when running driver tests

camsaul commented on test/metabase/test/data/generic_sql.clj:
```diff
@@ -248,15 +248,19 @@
+(defn- jdbc-execute! [db-spec sql]
+  (jdbc/execute! db-spec [sql] {:transaction? false, :multi? true}))
+
+(defn default-execute-sql! [driver context dbdef sql & {:keys [execute]
```
this param should probably be called `execute!` to make it clear that it's a function with side-effects 

https://github.com/bbatsov/clojure-style-guide#changing-state-fns-with-exclamation-mark

camsaul commented on test/metabase/test/data/generic_sql.clj:
```diff
@@ -294,11 +297,11 @@
   Since there are some cases were you might want to execute compound statements without splitting, an upside-down
   ampersand (`⅋`) is understood as an \"escaped\" semicolon in the resulting SQL statement."
-  [driver context dbdef sql]
+  [driver context dbdef sql  & {:keys [execute] :or {execute default-execute-sql!}}]
```
`execute` -> `execute!`

camsaul commented:
New PR is here: #7323

camsaul commented:
Unfortunately the table aliasing stuff breaks some of our other drivers e.g. BigQuery

```
"Field 'test_data.checkins.id' not found in table 't1'.
```

wjoel commented:
The BigQuery driver does some interesting things to the generated SQL, but BigQuery does support table aliases so it should be fairly easy to fix that.

camsaul commented:
I just merged PR #7323 which is based on this one.

AnonyV commented:
Hi there,
im new in Clojure,
Unfortunately, this [SparkSQL driver](https://s3.amazonaws.com/sparksql-deps/metabase-sparksql-deps-1.2.1.spark2-standalone.jar) doesn't work with my hiveserver.
I found that only jdbc:hive but not jdbc:hive2 would be worked.
so maybe the root cause is "org.apache.hive.jdbc", how can i use "[org.apache.hadoop.hive.jdbc](https://mvnrepository.com/artifact/org.apache.hadoop.hive/hive-jdbc)" instead?
any idea of this?

AnonyV commented:
> Hi there,
> im new in Clojure,
> Unfortunately, this [SparkSQL driver](https://s3.amazonaws.com/sparksql-deps/metabase-sparksql-deps-1.2.1.spark2-standalone.jar) doesn't work with my hiveserver.
> I found that only jdbc:hive but not jdbc:hive2 would be worked.
> so maybe the root cause is "org.apache.hive.jdbc", how can i use "[org.apache.hadoop.hive.jdbc](https://mvnrepository.com/artifact/org.apache.hadoop.hive/hive-jdbc)" instead?
> any idea of this?

OR if there can support hive and hive2 like this:
![qq 20181219183209](https://user-images.githubusercontent.com/35250120/50215600-7c901e00-03be-11e9-9e13-e57800b4ea2c.png)
